libs/ml/include/ml/optimisation/adagrad_optimiser.hpp

```diff
@@ -0,0 +1,101 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "math/standard_functions/sqrt.hpp"
+#include "ml/graph.hpp"
+#include "ml/ops/loss_functions/criterion.hpp"
+#include "ml/optimisation/optimiser.hpp"
+
+namespace fetch {
+namespace ml {
+namespace optimisers {
+
+/**
+ * Adaptive Gradient Algorithm optimiser
+ * i.e. Modified stochastic gradient descent with per-parameter learning rate
+ * @tparam T ArrayType
+ * @tparam C CriterionType
+ */
+template <class T, class C>
+class AdaGradOptimiser : public Optimiser<T, C>
```

Comment:
btw I know these optimisers classes are quite short (100 lines); but I would still like them written with only declarations in the class body and function definitions below. It's still a lot easier to read for most developers if they look and see immediately "okay this class just has a constructor an ApplyGradients(), and a ResetCache(). Plus it maintains consistency with the larger classes